File: droid/app/src/main/java/com/storyteller/droid/model/ModelDownloadManager.kt
package com.storyteller.droid.model

import android.os.StatFs
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import okhttp3.Call
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.security.MessageDigest

sealed interface ModelDownloadState {
    data object NotInstalled : ModelDownloadState
    data class Downloading(val downloadedBytes: Long, val totalBytes: Long) : ModelDownloadState
    data object Verifying : ModelDownloadState
    data object Installed : ModelDownloadState
    data object Cancelled : ModelDownloadState
    data class Failed(val code: String, val detail: String) : ModelDownloadState
}

/** Resumable, verified and atomically publishing model downloader. */
class ModelDownloadManager(
    private val modelsDir: File,
    private val client: OkHttpClient,
    private val spec: ReleaseModelSpec = ReleaseModelRegistry.gameMaster,
    private val downloadUrl: String = spec.downloadUrl,
    private val availableBytesProvider: (File) -> Long = ::availableBytes,
) {
    private val mutableState = MutableStateFlow<ModelDownloadState>(initialState())
    val state: StateFlow<ModelDownloadState> = mutableState.asStateFlow()
    @Volatile private var activeCall: Call? = null
    @Volatile private var cancelled = false

    val installedFile: File get() = File(modelsDir, spec.filename)
    val partialFile: File get() = File(modelsDir, "${spec.filename}.part")
    val isInstalled: Boolean get() = installedFile.isFile && installedFile.length() == spec.byteSize

    /** Must only be called after the user explicitly accepts the displayed model terms. */
    @Synchronized
    fun downloadAfterConsent() {
        if (state.value is ModelDownloadState.Downloading || state.value is ModelDownloadState.Verifying) return
        if (isInstalled) {
            mutableState.value = ModelDownloadState.Installed
            return
        }
        cancelled = false
        modelsDir.mkdirs()
        val existing = partialFile.takeIf { it.isFile }?.length() ?: 0L
        val remaining = spec.byteSize - existing
        if (remaining <= 0L || availableBytesProvider(modelsDir) < remaining) {
            mutableState.value = ModelDownloadState.Failed("MODEL_INSUFFICIENT_STORAGE", "At least $remaining additional bytes are required")
            return
        }

        val requestBuilder = Request.Builder().url(downloadUrl)
        if (existing > 0L) requestBuilder.header("Range", "bytes=$existing-")
        val call = client.newCall(requestBuilder.build())
        activeCall = call
        try {
            call.execute().use { response ->
                val validStatus = response.code == 200 && existing == 0L || response.code == 206 && existing > 0L
                if (!validStatus) throw DownloadFailure("MODEL_RESUME_REJECTED", "Unexpected HTTP ${response.code} for offset $existing")
                if (existing > 0L && response.header("Content-Range")?.startsWith("bytes $existing-") != true) {
                    throw DownloadFailure("MODEL_RESUME_REJECTED", "Invalid Content-Range for offset $existing")
                }
                val body = response.body ?: throw DownloadFailure("MODEL_EMPTY_RESPONSE", "Response has no body")
                var downloaded = existing
                FileOutputStream(partialFile, existing > 0L).buffered().use { output ->
                    body.byteStream().use { input ->
                        val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
                        while (true) {
                            if (cancelled) throw DownloadCancelled()
                            val count = input.read(buffer)
                            if (count < 0) break
                            output.write(buffer, 0, count)
                            downloaded += count
                            mutableState.value = ModelDownloadState.Downloading(downloaded, spec.byteSize)
                        }
                    }
                }
            }
            if (partialFile.length() != spec.byteSize) throw DownloadFailure("MODEL_SIZE_MISMATCH", "Expected ${spec.byteSize}, got ${partialFile.length()}")
            mutableState.value = ModelDownloadState.Verifying
            if (sha256(partialFile) != spec.sha256) {
                partialFile.delete()
                throw DownloadFailure("MODEL_CHECKSUM_MISMATCH", "Downloaded model failed SHA-256 verification")
            }
            publishAtomically(partialFile, installedFile)
            mutableState.value = ModelDownloadState.Installed
        } catch (_: DownloadCancelled) {
            mutableState.value = ModelDownloadState.Cancelled
        } catch (error: DownloadFailure) {
            mutableState.value = ModelDownloadState.Failed(error.code, error.message ?: error.code)
        } catch (error: IOException) {
            mutableState.value = if (cancelled) ModelDownloadState.Cancelled else ModelDownloadState.Failed("MODEL_NETWORK_ERROR", error.message ?: "Network error")
        } finally {
            activeCall = null
        }
    }

    fun cancel() {
        cancelled = true
        activeCall?.cancel()
    }

    @Synchronized
    fun deleteInstalledAndPartial() {
        cancel()
        installedFile.delete()
        partialFile.delete()
        mutableState.value = ModelDownloadState.NotInstalled
    }

    private fun initialState(): ModelDownloadState = if (isInstalled) ModelDownloadState.Installed else ModelDownloadState.NotInstalled

    private class DownloadFailure(val code: String, message: String) : IOException(message)
    private class DownloadCancelled : IOException()

    companion object {
        fun availableBytes(directory: File): Long = StatFs(directory.absolutePath).availableBytes

        fun sha256(file: File): String {
            val digest = MessageDigest.getInstance("SHA-256")
            file.inputStream().buffered().use { input ->
                val buffer = ByteArray(1024 * 1024)
                while (true) {
                    val count = input.read(buffer)
                    if (count < 0) break
                    digest.update(buffer, 0, count)
                }
            }
            return digest.digest().joinToString("") { "%02x".format(it) }
        }

        fun publishAtomically(partial: File, destination: File) {
            try {
                Files.move(partial.toPath(), destination.toPath(), StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING)
            } catch (_: java.nio.file.AtomicMoveNotSupportedException) {
                Files.move(partial.toPath(), destination.toPath(), StandardCopyOption.REPLACE_EXISTING)
            }
        }
    }
}

File: ios/StoryTeller/Model/ModelDownloadManager.swift
import CryptoKit
import Foundation

enum ModelDownloadState: Equatable {
    case notInstalled
    case downloading(downloadedBytes: Int64, totalBytes: Int64)
    case verifying
    case installed
    case cancelled
    case failed(code: String, detail: String)
}

/// Resumable downloader which verifies the release digest before atomic publication.
@MainActor
final class ModelDownloadManager: ObservableObject {
    @Published private(set) var state: ModelDownloadState
    private(set) var task: Task<Void, Never>?
    let spec: ReleaseModelSpec
    let modelsDirectory: URL

    var installedURL: URL { modelsDirectory.appendingPathComponent(spec.filename) }
    var partialURL: URL { modelsDirectory.appendingPathComponent("\(spec.filename).part") }
    var isInstalled: Bool { Self.fileSize(installedURL) == spec.byteSize }

    init(modelsDirectory: URL, spec: ReleaseModelSpec = ReleaseModelRegistry.gameMaster) {
        self.modelsDirectory = modelsDirectory
        self.spec = spec
        state = Self.fileSize(modelsDirectory.appendingPathComponent(spec.filename)) == spec.byteSize ? .installed : .notInstalled
    }

    /// The caller must present the registry source/license and obtain explicit consent first.
    func downloadAfterConsent() {
        guard task == nil, !isInstalled else {
            state = .installed
            return
        }
        task = Task { [weak self] in
            guard let self else { return }
            defer { task = nil }
            do {
                try await performDownload()
            } catch is CancellationError {
                state = .cancelled
            } catch let failure as ModelDownloadFailure {
                state = .failed(code: failure.code, detail: failure.detail)
            } catch {
                state = Task.isCancelled
                    ? .cancelled
                    : .failed(code: "MODEL_NETWORK_ERROR", detail: error.localizedDescription)
            }
        }
    }

    func cancel() {
        task?.cancel()
    }

    func deleteInstalledAndPartial() throws {
        cancel()
        for url in [installedURL, partialURL] where FileManager.default.fileExists(atPath: url.path) {
            try FileManager.default.removeItem(at: url)
        }
        state = .notInstalled
    }

    private func performDownload() async throws {
        let fileManager = FileManager.default
        try fileManager.createDirectory(at: modelsDirectory, withIntermediateDirectories: true)
        let existing = Self.fileSize(partialURL) ?? 0
        let remaining = spec.byteSize - existing
        guard remaining > 0 else {
            throw ModelDownloadFailure(code: "MODEL_SIZE_MISMATCH", detail: "Partial file is not resumable")
        }
        guard try availableBytes() >= remaining else {
            throw ModelDownloadFailure(code: "MODEL_INSUFFICIENT_STORAGE", detail: "At least \(remaining) additional bytes are required")
        }

        var request = URLRequest(url: spec.downloadURL)
        if existing > 0 { request.setValue("bytes=\(existing)-", forHTTPHeaderField: "Range") }
        let (bytes, response) = try await URLSession.shared.bytes(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ModelDownloadFailure(code: "MODEL_INVALID_RESPONSE", detail: "Download did not return HTTP")
        }
        let validStatus = (existing == 0 && http.statusCode == 200) || (existing > 0 && http.statusCode == 206)
        guard validStatus else {
            throw ModelDownloadFailure(code: "MODEL_RESUME_REJECTED", detail: "Unexpected HTTP \(http.statusCode) for offset \(existing)")
        }
        if existing > 0 && !(http.value(forHTTPHeaderField: "Content-Range")?.hasPrefix("bytes \(existing)-") ?? false) {
            throw ModelDownloadFailure(code: "MODEL_RESUME_REJECTED", detail: "Invalid Content-Range for offset \(existing)")
        }

        if existing == 0 { fileManager.createFile(atPath: partialURL.path, contents: nil) }
        let handle = try FileHandle(forWritingTo: partialURL)
        defer { try? handle.close() }
        try handle.seekToEnd()
        var downloaded = existing
        var buffer = Data()
        buffer.reserveCapacity(1024 * 1024)
        for try await byte in bytes {
            try Task.checkCancellation()
            buffer.append(byte)
            if buffer.count >= 1024 * 1024 {
                try handle.write(contentsOf: buffer)
                downloaded += Int64(buffer.count)
                buffer.removeAll(keepingCapacity: true)
                state = .downloading(downloadedBytes: downloaded, totalBytes: spec.byteSize)
            }
        }
        if !buffer.isEmpty {
            try handle.write(contentsOf: buffer)
            downloaded += Int64(buffer.count)
        }
        try handle.synchronize()
        guard downloaded == spec.byteSize else {
            throw ModelDownloadFailure(code: "MODEL_SIZE_MISMATCH", detail: "Expected \(spec.byteSize), got \(downloaded)")
        }

        state = .verifying
        guard try Self.sha256(partialURL) == spec.sha256 else {
            try? fileManager.removeItem(at: partialURL)
            throw ModelDownloadFailure(code: "MODEL_CHECKSUM_MISMATCH", detail: "Downloaded model failed SHA-256 verification")
        }
        try Self.publishAtomically(partialURL, to: installedURL)
        state = .installed
    }

    private func availableBytes() throws -> Int64 {
        let values = try modelsDirectory.resourceValues(forKeys: [.volumeAvailableCapacityForImportantUsageKey])
        return values.volumeAvailableCapacityForImportantUsage ?? 0
    }

    nonisolated static func fileSize(_ url: URL) -> Int64? {
        let attributes = try? FileManager.default.attributesOfItem(atPath: url.path)
        return (attributes?[.size] as? NSNumber)?.int64Value
    }

    nonisolated static func sha256(_ url: URL) throws -> String {
        let handle = try FileHandle(forReadingFrom: url)
        defer { try? handle.close() }
        var digest = SHA256()
        while let data = try handle.read(upToCount: 1024 * 1024), !data.isEmpty { digest.update(data: data) }
        return digest.finalize().map { String(format: "%02x", $0) }.joined()
    }

    nonisolated static func publishAtomically(_ partial: URL, to destination: URL) throws {
        let fileManager = FileManager.default
        if fileManager.fileExists(atPath: destination.path) {
            _ = try fileManager.replaceItemAt(destination, withItemAt: partial)
        } else {
            try fileManager.moveItem(at: partial, to: destination)
        }
    }
}

private struct ModelDownloadFailure: Error {
    let code: String
    let detail: String
}
